fix: PR creation from submodules

ddorando opened the pull request:
This fixes issues you can encounter when you attempt to create a PR for a checked-out submodule.

- [x] Use 'git rev-parse --git-path' instead of 'git rev-parse --git-dir' to find repository config file
The previous code assumes that the git directory (git rev-parse --git-dir) is always a relative path. It will break if the path returned by this command is an absolute one.
By using 'git rev-parse --git-path <path>' only, instead of manually piecing together the path for the config file, we pretty much avoid this problem.
- [x] Use 'git rev-parse --is-shallow-repository' instead of the check whether the repository is shallow by checking whether a path in the .git directory exists (.git/shallow)

peter-evans commented:
Hi @ddorando 

Thanks for contributing! There are a couple of tests that are failing. Please take a look.

ddorando commented on src/utils.ts:
```diff
@@ -149,7 +149,7 @@ export function fileExistsSync(path: string): boolean {
   let stats: fs.Stats
   try {
-    stats = fs.statSync(path)
+    stats = fs.statSync(fs.realpathSync(path))
```
Remainder from development. Removing.

peter-evans commented:
/test repository=ddorando/create-pull-request ref=main build=true

ddorando commented:
I've reverted my last commit ee18c00139ac748b9142cb62ce9a2aa8fec9561a. It turns out that the body-path passed in to the action is the path to this .git/shallow file, which in turn gets stat' in the fileExistsSync function, which only gets called [here](https://github.com/peter-evans/create-pull-request/blob/main/src/create-pull-request.ts#L43), other than the call I've removed. It appears to somehow fail following the symlink, and thus we get this ENOTDIR.

Honestly, it's your call, but I'd suggest to deal that in an issue dedicated for that instead of handling that issue in this PR.

If you'd like to see the logs runtime logs for both cases, I can forward them to you. I'll also attempt to make a minimum failing example for it.

ddorando commented:
Minimum "working" example: https://github.com/ddorando/create-pr-example
I know it doesn't create a PR for the "works" pipeline - I am a bit too lazy to fully set it up, though you should see the difference there.

peter-evans commented:
I've decided to add a test for PR creation from submodules to my [test suite](https://github.com/peter-evans/create-pull-request-tests) since this is the kind of thing that could easily get broken with changes in future. I think it will help me to understand the problem and what you have done to solve it.

peter-evans commented:
I added a test here: https://github.com/peter-evans/create-pull-request-tests/blob/master/.github/workflows/test-command.yml#L1068-L1087

Then I ran the tests on the latest version of the action and it failed on the new test, as you can see here: https://github.com/peter-evans/create-pull-request-tests/actions/runs/6160153160/job/16716491503

I also ran it against this PR and it still failed in the same place: https://github.com/peter-evans/create-pull-request-tests/actions/runs/6160279137/job/16716982228

So I'm bit confused about what this PR is attempting to fix. I'm also wondering if what you are trying to do is even possible. I don't know that it's possible to create a PR with changes for submodules. I'm fairly sure that it wouldn't work because you need to commit to the submodule linked repo before the commit hash change is available to commit for the PR.

Maybe I'm misunderstanding what you are trying to do. If that's the case please explain so I can create a failing test in my suite that can pass with your changes.

ddorando commented:
So this PR only attempts to fix making a PR from within the context of a submodule, so a test that would work in this case would be something along the lines of:
```yaml
  testCreateFromWithinSubmodules:
    needs: beforeTest
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
        with:
          submodules: true
      # Download the action artifact
      - uses: actions/download-artifact@v3
        with:
          name: create-pull-request
          path: create-pull-request-local
      # Test
      - name: Create report file
        run: date +%s > create-pull-request-tests-submodule/report.txt
      - name: Create Pull Request
        uses: ./create-pull-request-local
        with:
          title: testCreateWithSubmodules
          branch: tests/create-from-within-submodules
          path: create-pull-request-tests-submodule
```
(so pretty much the only real change in the test you have right now would be the last line I added)

Also, the workflow I currently have (which is also how I tested my changes here) is as following:
- Clone of a project with a submodule
- Run dependency updates for the entire project, including the submodule
- Make a PR (1) on the submodule that updates the deps for it
- Make a PR (2) on the project itself for the dep updates. The submodule reference won't get updated here (and it won't error out git, so this would be fine).
- (manual step) Bump the submodule reference in the PR (2) on the project repo once the submodule's PR (1) is merged in.

Hope this clears things up.

peter-evans commented:
@ddorando Thank you, that cleared things up. I understand what you are trying to do now. However, I think you approach is a bit unnecessary. You can use this action to checkout the submodule repository directly and raise a PR. There is no need to do it inside the parent repository. See the docs here about creating a PR in a remote repository: https://github.com/peter-evans/create-pull-request/blob/main/docs/concepts-guidelines.md#creating-pull-requests-in-a-remote-repository

I'm assuming that you want the workflow to run in the parent repository, hence the above linked example in the docs. The other alternative, of course, is to run the workflow to raise a PR in the submodule repository itself.

As for what you are attempting to do with the changes in this PR. It seems to work to a point, but then runs into a [permissions error](https://github.com/peter-evans/create-pull-request-tests/actions/runs/6174904384/job/16760685230). This is because the token in use doesn't have write access to the submodule repository. This could probably be worked around, but I don't think this approach makes sense when there are other much more simple approaches to achieve the same result.
